lib/connectionProvider.js

```diff
@@ -0,0 +1,25 @@
+const { isFunction, cloneDeep } = require('lodash');
+
+class ConnectionConfigProvider {
+  constructor(client) {
+    this.client = client;
+    this.resolvedConnectionSettings = null;
+  }
+
+  async resolveConnectionConfig() {
+    if (this.resolvedConnectionSettings) return this.resolvedConnectionSettings;
+    if (isFunction(this.client.config.connection)) {
```

Comment:
We can do the check once in constructor and set some flag after that.

Reply:
Can you give more details about this, please?

if we need to manage this in `ConnectionConfigProvider` constructor, the how-to manipulate the promise as the type of the function is async?

Reply:
What I mean is this:

1) In constructor check if it's a function. If it is - set `this.isConfigSourceFunction = true` and do nothing else. If it is not - set `this.isConfigSourceFunction = false` and `this.resolvedConnectionSettings = cloneDeep <...>`

2) in `resolveConnectionConfig ` check if `this.resolvedConnectionSettings` is set, if it is, return, if it is not and `isConfigSourceFunction`, resolve from function, if neither, probably throw an error.

Reply:
I think this line in `ConnectionConfigProvider` class does the same, right?
```
  async resolveConnectionConfig() {
    if (this.resolvedConnectionSettings) return this.resolvedConnectionSettings;
```
the lines after check for the type of the connection. if it is a function resolve it, other wise, deepclone it.

Reply:
Partially. What I'm saying is that everything that can be moved from `resolveConnectionConfig ` to constructor should ideally be moved, `resolveConnectionConfig ` should be as simple as possible.
Current implementation is different from what I've described.